### ISSUE TITLE
Incorrect cropping on graph

### ISSUE THREAD

**Issue by sean-fitzpatrick** [APEXCalculus/APEXCalculusPTX#261]
See the graph in [this exercise](http://localhost:8000/sec_limits_infty.html#ex-limit-infinity-use-graph-2). The TikZ for this has the same bounds as other exercises in this set, but for some reason, cropping is leaving a bunch of whitespace at the bottom.

I can't spot anything in the TikZ code that explains this.

Here's the code:

```
<latex-image>
  \begin{tikzpicture}
    \begin{axis}[
      xmin = $xmin,
      xmax = $xmax,
      ymin = -50,
      ymax = 50,
    ]
      \addplot[firstcurvestyle, -{Kite}, domain=$xmin:$x[0], samples=100] {1/((x-$a[0])*(x-$a[1])^(2))};
      \addplot[firstcurvestyle, {Kite}-{Kite}, domain=$x[1]:$x[2], samples=100] {1/((x-$a[0])*(x-$a[1])^(2))};
      \addplot[firstcurvestyle, {Kite}-, domain=$x[3]:$xmax, samples=100] {1/((x-$a[0])*(x-$a[1])^(2))};
      \addplot[asymptote,-] coordinates {($a[0],-50) ($a[0],50)};
      \addplot[asymptote,-] coordinates {($a[1],-50) ($a[1],50)};
    \end{axis}
  \end{tikzpicture}
</latex-image>
```

**Comment by APEXCalculus** [comment]
The included link starts with 'localhost' so I can't see what you see.

I substituted values for $xmin, etc., to see what the graphic would create. It looks fine to me. 

I'm assuming that x[0] is a value similar to a[0]-0.01, and x[1] is about a[0]+0.01, etc.? Then the domains make sense. Maybe the error is in there somewhere.

**Comment by sean-fitzpatrick** [comment]
Oh, shoot! Sorry, I usually know better than to do that.
The version [currently online](https://opentext.uleth.ca/apex-calculus/sec_limits_infty.html#exercise-166) seems OK.

The build I just did uses different x values for the two asymptotes, but the graph looks good, except for the fact that it didn't get cropped correctly.

**Comment by sean-fitzpatrick** [comment]
You can see the issue on the [Runestone version](https://runestone.academy/ns/books/published/APEX/sec_limits_infty.html?mode=browsing#ex-limit-infinity-use-graph-2)

**Comment by APEXCalculus** [comment]
By the fact you can't see the arrow heads, I wonder if this is an overflow type issue where the dimensions got too big. I can recreate the effect a bit on my own, but LaTeX also throws an error. I'm assuming you are being told that all is ok, but you still see the cropping issue.

My recommendation would be to restrict the domain farther from the asymptotes.

**Comment by sean-fitzpatrick** [comment]
Ah, good catch. I guess it depends on where the asymptotes are, since it's OK for some versions. I'll push things away from the asymptotes a little.

**Comment by sean-fitzpatrick** [comment]
I've adjust things so that they look good for the version that's currently being included in the static version (the PDF, and HTML before the question is activated). But if questions are added later on, what looks good now might not look good later.
@Alex-Jordan is on vacation right now but I will probably need to defer to him.

Right now the webwork code chooses the two x values for the vertical asymptotes randomly between 1 and 9.
One asymptote comes from an `(x-a)` term, and the other from an `(x-b)^2` term. Sometimes a<b, and sometimes a>b.
I think the adjustment needed in one case is different from what is needed in the other case.

One solution could be to choose `a` randomly from 1 to 5, and let `b=a+c`, where c is chosen randomly between 1 and 3.

**Comment by APEXCalculus** [comment]
I think your latter solution is the most foolproof going forward. I haven't
done the small numerical analysis to figure out the difference between a<b
and b<a, but I bet that's an issue. You could make things a bit more random
by letting the numerator be either 1 or -1; that wouldn't throw off the
scale but would make things a bit different.

**Comment by sean-fitzpatrick** [comment]
I'm a bit confused as to why the existing code doesn't work as intended. It should! Something is getting mixed up.

The graph is y = 1/[(x-a)(x-b)^2]. We want the y range to be -50 to 50.

If we're near the x=a asymptote, set x=a+t. If we want to get to exactly y=50, then:

50 = 1/(t(b-a)^2), so t = 1/(50(b-a)^2).

If we're near the x=b asymptote, set x=b+t, and y=50, and we get

50 = 1/[(b-a)t^2], so t = 1/sqrt(50|b-a|)

This is exactly how the program is coded, so I'm not really sure how we're ending up with y values outside the -50 to 50 range.

**Comment by sean-fitzpatrick** [comment]
Oh, I see the mistake. If I put x=a+t, I get y = 1/[t(t+a-b)^2] -- we were ignoring the delta in the second factor!

**Comment by sean-fitzpatrick** [comment]
That means we are trying to solve a cubic equation...

**Comment by sean-fitzpatrick** [comment]
OK, I have a sort-of solution. Since I'm trying to solve a cubic, but the thing I'm solving for is small, I can approximate by dropping the cubic term, and then solve the resulting quadratic. Not perfect, but I tested with about 20 different randomizations, and the graph fits each time.

**Comment by APEXCalculus** [comment]
Ha! Gotta love numerical. Yes, the t^3 term will be small, and bob's your uncle. That's great!

**Comment by sean-fitzpatrick** [comment]
It doesn't quite solve the problem. The issue (even if you were to try to go with a cubic solution) is that the value to the left of the asymtpote will never match the value to the right: if we're a distance d from the asymptote, you have (d+(a-b)) on one side, and (-d+(a-b)) on the other. One of these is larger in absolute value than the other, depending on the sign of a-b.

Now, you could use different distances on either side of the asymptote, but if you want the same y (absolute) value on either side, the pair of distances ends up needing to be a point on a hyperbola! You need:

(d_1 + (a-b)/2)^2 - (d_2 - (a-b)/2)^2 = (a-b)^2/2

Somewhere in here is a lesson plan for our students.

**Comment by sean-fitzpatrick** [comment]
I was finding a few seeds where one part of the graph would still run past the y bound.
Even though TikZ doesn't display the part of the graph that exceeds the y bound, there is some vestige of it that the pdf crop utility sees, leading to unwanted whitespace at the top or bottom of the image.

The other issue is a minor aesthetic thing: you'd like the arrow tips on either side of the asymptote to be at the same height, but they're always slightly off.

In any case, I think I have a solution that works. I had to insist that the asymptote from the quadratic term is always to the right of the other. That way I know the sign of a-b, and I can comensate by different amounts on either side. Those amounts weren't mathematically determined. More like scientifically: some trial and error, and then checking through lots of random seeds. There are only 10 different versions now (there used to be around 3 dozen, I think) but they all look OK.

The number of versions is down also because I restricted the gap between the asymptotes to be 1 or 2 units, and kept the maximum x value below 5. I found that when the asymptotes were too far apart, the graph didn't look good at scale: it would sit almost on the x axis until right next to the asymptote, and then take a sharp turn upwards.
